WebVTT regions are not positioned according to the spec

danae404 opened the pull request:
#### eae66f440e600b0d39e9bcbd60582c96cc4c35a0
<pre>
WebVTT regions are not positioned according to the spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=260402">https://bugs.webkit.org/show_bug.cgi?id=260402</a>
rdar://23091897

Reviewed by Eric Carlson.

Updating WebVTT code to reflect changes that have been made to the W3C WebVTT spec,
specifically in section 7 (<a href="https://www.w3.org/TR/webvtt1/#processing-model).">https://www.w3.org/TR/webvtt1/#processing-model).</a>
In step 14.2 in section 7.1, there is a split in logic depending on if the cue
Is within a region or not. Currently, the code always follows the path for
Cues without regions. I updated this so that step 14.3 of the spec would be
executed for cues assigned to regions. I also updated the CSS applied to the
region and text track display divs to conform to section 7.4 &quot;Applying CSS
Properties to WebVTT Node Objects&quot;.

* LayoutTests/media/track/captions-webvtt/regions.vtt: Added.
* LayoutTests/media/track/webvtt-region-position-expected.txt: Added.
* LayoutTests/media/track/webvtt-region-position.html: Added.
* Source/WebCore/Modules/modern-media-controls/controls/text-tracks.css:
(::-webkit-media-text-track-display):
(::-webkit-media-text-track-region):
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateDisplay):
(WebCore::MediaControlTextTrackContainerElement::processActiveVTTCue):
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCueBox::applyCSSPropertiesWithRegion):
(WebCore::VTTCueBox::applyCSSProperties):
(WebCore::VTTCue::calculateDisplayParametersWithRegion):
(WebCore::VTTCue::obtainCSSBoxes):
(WebCore::VTTCue::getDisplayTree):
* Source/WebCore/html/track/VTTCue.h:
* Source/WebCore/html/track/VTTRegion.cpp:
(WebCore::VTTRegion::prepareRegionDisplayTree):

Canonical link: <a href="https://commits.webkit.org/268787@main">https://commits.webkit.org/268787@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6484e7704efcef67b54733b983bafd39695138c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22454 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19193 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20793 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21160 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20569 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20779 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20628 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17877 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23312 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17795 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18682 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24960 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18873 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18859 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22898 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19441 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16523 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18661 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18518 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4962 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22997 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19260 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/Modules/modern-media-controls/controls/text-tracks.css:
```diff
@@ -76,6 +77,20 @@ video[controls]::-webkit-media-text-track-container.visible-controls-bar {
+::-webkit-media-text-track-region {
+    position: absolute;
+    writing-mode: horizontal-tb;
+    background: rgna(0,0,0,0.8);
```
should this be `rgba` instead of `rgna`?

danae404 replied:
Yes!

aestes commented on Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
```diff
@@ -217,11 +225,13 @@ void MediaControlTextTrackContainerElement::processActiveVTTCue(VTTCue& cue)
             appendChild(region->getDisplayTree());
 
+        Ref<TextTrackCueBox> displayBox = *cue.getDisplayTree();
         region->appendTextTrackCueBox(WTFMove(displayBox));
```
```suggestion
        region->appendTextTrackCueBox(*cue.getDisplayTree());
```

aestes commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -769,6 +788,34 @@ double VTTCue::calculateMaximumSize() const
+    // Subtract half of region’s WebVTT region width from offset.
+    auto computedPositionAlignment = calculateComputedPositionAlignment();
+    if (computedPositionAlignment == PositionAlignSetting::Center)
+        offset = offset - (regionWidth / 2);
```
```suggestion
        offset = offset - regionWidth / 2;
```

eric-carlson commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -148,6 +148,25 @@ VTTCueBox::VTTCueBox(Document& document, VTTCue& cue)
+{
+    auto textTrackCue = getCue();
+    ASSERT(!textTrackCue || is<VTTCue>(textTrackCue));
+    if (!is<VTTCue>(textTrackCue))
```
The `ASSERT` and the `if` for an early return should test the same conditions

danae404 replied:
Is there a preference for whether to keep the `ASSERT` or the `if`?

eric-carlson replied:
An ASSERT is a noop in a release build, so if both conditions are unexpected we want to check both in the assert in a debug build so we can find and fix the code, and check both and return early to avoid crashing in a release build.

danae404 replied:
That makes sense!

eric-carlson commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -769,6 +788,34 @@ double VTTCue::calculateMaximumSize() const
 }
 
+void VTTCue::calculateDisplayParametersWithRegion()
+{
+    // 1. Let region be cue’s WebVTT cue region.
```
It is probably worth adding an `ASSERT(region())`

eric-carlson commented on LayoutTests/media/track/opera/track/webvtt/parsing-cue-data/tests/entities-expected.txt:
```diff
@@ -0,0 +1,3 @@
+FAIL: Timed out waiting for notifyDone to be called
+
+
```
Did you mean to include this file?

eric-carlson commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -769,6 +788,35 @@ double VTTCue::calculateMaximumSize() const
+    // Subtract half of region’s WebVTT region width from offset.
+    auto computedPositionAlignment = calculateComputedPositionAlignment();
+    if (computedPositionAlignment == PositionAlignSetting::Center)
+        offset = offset - regionWidth / 2;
```
Nit: `offset - (regionWidth / 2)` would be clearer

aestes commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -148,6 +148,29 @@ VTTCueBox::VTTCueBox(Document& document, VTTCue& cue)
+    Ref cue = downcast<VTTCue>(*textTrackCue);
+
+    // the 'left' property must be set to left
+    std::visit(WTF::makeVisitor([&] (double left) {
```
Nit: may be a little simpler to use WTF::switchOn rather than std::visit + WTF::makeVisitor.

danae404 committed:
WebVTT regions are not positioned according to the spec
https://bugs.webkit.org/show_bug.cgi?id=260402
rdar://23091897

Reviewed by Eric Carlson.

Updating WebVTT code to reflect changes that have been made to the W3C WebVTT spec,
specifically in section 7 (https://www.w3.org/TR/webvtt1/#processing-model).
In step 14.2 in section 7.1, there is a split in logic depending on if the cue
Is within a region or not. Currently, the code always follows the path for
Cues without regions. I updated this so that step 14.3 of the spec would be
executed for cues assigned to regions. I also updated the CSS applied to the
region and text track display divs to conform to section 7.4 "Applying CSS
Properties to WebVTT Node Objects".

* LayoutTests/media/track/captions-webvtt/regions.vtt: Added.
* LayoutTests/media/track/webvtt-region-position-expected.txt: Added.
* LayoutTests/media/track/webvtt-region-position.html: Added.
* Source/WebCore/Modules/modern-media-controls/controls/text-tracks.css:
(::-webkit-media-text-track-display):
(::-webkit-media-text-track-region):
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateDisplay):
(WebCore::MediaControlTextTrackContainerElement::processActiveVTTCue):
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCueBox::applyCSSPropertiesWithRegion):
(WebCore::VTTCueBox::applyCSSProperties):
(WebCore::VTTCue::calculateDisplayParametersWithRegion):
(WebCore::VTTCue::obtainCSSBoxes):
(WebCore::VTTCue::getDisplayTree):
* Source/WebCore/html/track/VTTCue.h:
* Source/WebCore/html/track/VTTRegion.cpp:
(WebCore::VTTRegion::prepareRegionDisplayTree):

Canonical link: https://commits.webkit.org/268787@main
